eipw-lint/src/lints/markdown/link_eip.rs

```diff
@@ -0,0 +1,141 @@
+        let pattern = r"(?i)\b(?:eip|erc)-(\d+)\b";
+        let url_re = Regex::new(pattern).map_err(Error::custom)?;
+
+        let url_eip_number = if let Some(captures) = url_re.captures(&self.current_link.url) {
+            captures.get(1).map(|m| m.as_str())
+        } else { None };
+
+        if let Some(url_eip_number) = url_eip_number {
+            let section_pattern = r"eip-([^.]*)\.md#.+$";
+            let url_re = Regex::new(section_pattern).map_err(Error::custom)?;
+            let dynamic_pattern = if url_re.is_match(&self.current_link.url) {
+                format!(r"^(EIP|ERC)-{}\s*\S+", regex::escape(&url_eip_number))
+            } else {
+                format!(r"^(EIP|ERC)-{}$", regex::escape(&url_eip_number))
+            };
+            let text_re = Regex::new(&dynamic_pattern).map_err(Error::custom)?;
+        
+            if text_re.is_match(&self.current_link.text) {
+                return Ok(Next::TraverseChildren);
+            };
+        
+            let expected = if url_re.is_match(&self.current_link.url) {
+                format!("[EIP|ERC-{}<section-description>]", url_eip_number)
+            } else {
+                format!("[EIP|ERC-{}]", url_eip_number)
+            };
+        
+            let footer_label = format!("link text should match `{}`", expected);
+        
+            let source = self.ctx.source_for_text(ast.sourcepos.start.line, &self.current_link.text);
+            self.ctx.report(Snippet {
+                title: Some(Annotation {
+                    annotation_type: self.ctx.annotation_type(),
+                    id: Some(self.slug),
+                    label: Some("link text does not match link destination"),
+                }),
+                slices: vec![Slice {
+                    fold: false,
+                    line_start: ast.sourcepos.start.line,
+                    origin: self.ctx.origin(),
+                    source: &source,
+                    annotations: vec![],
+                }],
+                footer: vec![Annotation {
+                    id: None,
+                    annotation_type: AnnotationType::Info,
+                    label: Some(&footer_label),
+                }],
+                opt: Default::default(),
+            })?;
+        }
+
+        Ok(Next::TraverseChildren)
+    }
+}
+
+impl<'a, 'b, 'c> tree::Visitor for Visitor<'a, 'b, 'c> {
+    type Error = Error;
+
+    fn enter_link(&mut self, _: &Ast, link: &NodeLink,) -> Result<Next, Self::Error> {
+        if self.re.is_match(&link.url) {        
+            self.current_link = Link { url: link.url.to_owned(), text: String::new() }; 
+            self.link_depth += 1;  
+        }                          
+        Ok(Next::TraverseChildren)
+    }
+
+    fn depart_link(&mut self, _: &Ast, _: &NodeLink) -> Result<(), Self::Error> {
+        if self.link_depth > 0 {
+            self.link_depth = self.link_depth.checked_sub(1).unwrap();
+        }
```

Comment:
If `self.link_depth` is greater than zero, the `checked_sub` is unnecessary.

Reply:
`checked_sub` in the `depart_link` function returns self.link_depth to 0. I check link_depth in `enter_text` fn. If link_depth == 0, it means the text is not from the link, so, I skip it.

Reply:
What I mean is that if `self.link_depth > 0` then `self.link_depth - 1` can never underflow. The `checked_sub` will never fail.